kqueen/auth/ldap.py

```diff
@@ -22,31 +22,62 @@ def __init__(self, *args, **kwargs):
         self.connection = ldap.initialize(self.uri)
 
     @staticmethod
-    def _email_to_dn(email):
-        """This function reads email and converts it to LDAP dn
+    def _email_to_dn_dict(email):
+        """This function reads email and converts it to LDAP dn and dc's
 
         Args:
             email (str): e-mail address
 
         Returns:
-            dn (str): LDAP dn, like 'cn=admin,dc=example,dc=org'
+            dn (dict): LDAP dn and dc's, like {'cn': 'cn=admin', 'dcs': 'dc=example,dc=org'}
         """
 
-        segments = []
+        segments = {}
         if '@' in email:
             cn, dcs = email.split('@')
         else:
             cn = email
             dcs = ''
 
         if cn:
-            segments.append('cn={}'.format(cn))
+            segments['cn'] = cn
 
+        dc_list = []
         if '.' in dcs:
             for s in dcs.split('.'):
-                segments.append('dc={}'.format(s))
+                dc_list.append('dc={}'.format(s))
 
-        return ','.join(segments)
+        segments['dcs'] = ','.join(dc_list)
+        logger.debug('RegExp for search valid full-DN for Ldap user: {}'.format(segments))
+
+        return segments
+
+    def _get_matched_dn(self, username):
+        dn_segments = self._email_to_dn_dict(username)
+
+        if self._bind(self.kqueen_username, self.kqueen_password):
+            self.connection = ldap.initialize(self.uri)
```

Comment:
it should be already evaluated in init method, so you can just use self.connection right away

Reply:
no, previously we use unbind() only once, and all works fine
but in this case we use unbind multiple times, and each time we should re-init ldap object, otherwise if=t failed to bind. i can clarify info directly or take a look on LDAP docs